### ISSUE TITLE
Connection closed before response was received

### ISSUE THREAD

**Issue by Abouassi98** [cfug/dio#2170]
### Package

dio

### Version

5.4.2+1

### Operating-System

Android, iOS

### Adapter

Default Dio

### Output of `flutter doctor -v`

```shell
[✓] Flutter (Channel stable, 3.19.4, on macOS 14.4.1 23E224 darwin-arm64, locale en-EG)
[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
[✓] Xcode - develop for iOS and macOS (Xcode 15.3)
[✓] Chrome - develop for the web
[✓] Android Studio (version 2023.2)
[✓] VS Code (version 1.87.2)
[✓] Connected device (3 available)            
    ! Error: Browsing on the local area network for Me. Ensure the device is unlocked and attached with a cable or associated with the same
      local area network as this Mac.
      The device must be opted into Developer Mode to connect wirelessly. (code -27)
    ! Error: Browsing on the local area network for Mohamed’s Iphone (2). Ensure the device is unlocked and attached with a cable or
      associated with the same local area network as this Mac.
      The device must be opted into Developer Mode to connect wirelessly. (code -27)
    ! Error: Browsing on the local area network for ‏iPhone ‏Fadwa. Ensure the device is unlocked and attached with a cable or associated
      with the same local area network as this Mac.
      The device must be opted into Developer Mode to connect wirelessly. (code -27)
    ! Error: Browsing on the local area network for H. Sayed. Ensure the device is unlocked and attached with a cable or associated with the
      same local area network as this Mac.
      The device must be opted into Developer Mode to connect wirelessly. (code -27)
[✓] Network resources

• No issues found!
```

### Dart Version

3.3.2 

### Steps to Reproduce

1. Run with any version of Dio above 5.4.0.
2. try to send more requests in a short period of time. 

### Expected Result

When using Dio version 5.4.0, the connection is not closed and the response succeeds.

![untitled2](https://github.com/cfug/dio/assets/53384802/bbf873ac-5063-4297-a77f-cc6f2a095e71)

### Actual Result

flutter: The following exception was thrown getBOKRechargeRequestStatusProvider:
flutter: AppException.serverException(type: ServerExceptionType.unknown, message: HttpException: Connection closed before response was received, uri = https://api.bravo-sudan.com/bravo/api/v1/bok/getRechargeRequests, code: null)
flutter: 
#0      DioMixin.fetch (package:dio/src/dio_mixin.dart:509:7)
<asynchronous suspension>
#1      MainApiFacade._errorHandler (package:bravo/core/infrastructure/network/main_api/api_callers/main_api_facade.dart:66:14)
<asynchronous suspension>
#2      TransactionRemoteDataSource.getBOKRechargeRequestStatus (package:bravo/features/transaction/infrastructure/data_sourse/transaction_remote_remote_data_source.dart:48:20)
<asynchronous suspension>
#3      TransactionHistoryRepository.getBOKRechargeRequestStatus (package:bravo/features/transaction/infrastructure/repository/transaction_history_repository.dart:43:20)

![untitled](https://github.com/cfug/dio/assets/53384802/85f99108-b184-4291-bc29-ab1aa64e3b32)

**Comment by AlexV525** [comment]
If this is version specific, could you run `git bisect` to find out which commit breaks the functionality?

**Comment by Abouassi98** [comment]
I have just upgraded the Dio version from 5.4.0 to 5.4.2+1

**Comment by AlexV525** [comment]
Running bisect will help to determine which exact commit breaks the behavior.

**Comment by Abouassi98** [comment]
I am considering running a bisect to identify the problematic commit. However, I want to inform you that my recent commit only involved upgrading the version to 5.4.2+1

**Comment by AlexV525** [comment]
I understand your statement clearly, IMO. I'm asking you to do the bisect because you have a reproducible example and can debug the issue easily. Besides, simply switching between versions might not be helpful since every version includes a bunch of commits.

**Comment by Abouassi98** [comment]
I've completed running `git bisect` to identify the commit that introduced the issue in the Dio package. After testing multiple commits, I've found that the first bad bisect commit is "Fix streamed response cancellation and progress (#2068)".


https://github.com/cfug/dio/commit/845e06f4fde3537522b9dc813a5b6e0fdda4a4c5

**Comment by AlexV525** [comment]
> I've completed running `git bisect` to identify the commit that introduced the issue in the Dio package. After testing multiple commits, I've found that the first bad bisect commit is "Fix streamed response cancellation and progress (#2068)".
> 
> [845e06f](https://github.com/cfug/dio/commit/845e06f4fde3537522b9dc813a5b6e0fdda4a4c5)

Thanks for the bisect! Now could you split a minimal reproducible example so we can take a deeper investigation with your case?

**Comment by AlexV525** [comment]
cc @kuhnroyal to check the implementation as the code owner

**Comment by Abouassi98** [comment]
Sure! Here's the minimal reproducible example: [Minimal Reproducible Example](https://github.com/Abouassi98/test_dio).

**Comment by Abouassi98** [comment]
@AlexV525, do you have any updates there?

**Comment by AlexV525** [comment]
> @AlexV525, do you have any updates there?

We'll let you know if any updates. We generally works when we have spare time since this is an open-source project.

**Comment by kuhnroyal** [comment]
> Sure! Here's the minimal reproducible example: [Minimal Reproducible Example](https://github.com/Abouassi98/test_dio).

Thanks, sadly this is not minimal and requires to start and debug a flutter app. Can you reduce the actual request to a unit test or at least a simple Dart file?

**Comment by Abouassi98** [comment]
I'm certain this issue stems from using Riverpod, a common example in the docs.

```dart
final apiProvider = FutureProvider.autoDispose.family<List<dynamic>, String>((ref, endPoint) async {
  final cancelToken = CancelToken();
  ref.onDispose(() => cancelToken.cancel());
  final response = await ref.watch(dioProvider).get(endPoint, cancelToken: cancelToken);
  return response.data;
});
```
This issue happens with any version of Dio > 5.4.0, as I mentioned above.

And this is a Minimal Reproducible Example:

```dart
import 'package:dio/dio.dart';
import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

void main() {
  runApp(const ProviderScope(child: App()));
}

class App extends StatelessWidget {
  const App({super.key});

  @override
  Widget build(BuildContext context) => const MaterialApp(
        home: TabBarScreen(),
      );
}

class TabBarScreen extends StatefulWidget {
  const TabBarScreen({super.key});

  @override
  State<TabBarScreen> createState() => _TabBarScreenState();
}

class _TabBarScreenState extends State<TabBarScreen> with SingleTickerProviderStateMixin {
  late final TabController controller;
  @override
  void initState() {
    controller = TabController(
      length: 2,
      vsync: this,
    );

    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(),
      body: Column(
        mainAxisSize: MainAxisSize.min,
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          TabBar(
            indicator: const BoxDecoration(),
            padding: const EdgeInsets.all(8),
            controller: controller,
            tabs: const [
              Tab(
                text: 'Posts',
              ),
              Tab(
                text: 'Users',
              ),
            ],
          ),
          Expanded(
            child: TabBarView(
              controller: controller,
              children: const [
                TabBarViewItem(
                  key: PageStorageKey<String>('tap1'),
                  endPoint: 'posts',
                ),
                TabBarViewItem(
                  key: PageStorageKey<String>('tap2'),
                  endPoint: 'users',
                ),
              ],
            ),
          ),
        ],
      ),
    );
  }
}

final dioProvider = StateProvider<Dio>((ref) {
  return Dio()
    ..options = BaseOptions(
      baseUrl: 'https://gorest.co.in/public/v2/',
      connectTimeout: const Duration(seconds: 20),
      receiveTimeout: const Duration(seconds: 30),
    );
});

final apiProvider = FutureProvider.autoDispose.family<List<dynamic>, String>((ref, endPoint) async {
  final cancelToken = CancelToken();
  ref.onDispose(() => cancelToken.cancel());
  final response = await ref.watch(dioProvider).get(endPoint, cancelToken: cancelToken);
  return response.data;
});

class TabBarViewItem extends ConsumerWidget {
  const TabBarViewItem({
    super.key,
    required this.endPoint,
  });
  final String endPoint;
  @override
  Widget build(BuildContext context, WidgetRef ref) {
    final list = ref.watch(
      apiProvider(endPoint),
    );
    return Column(
      children: [
        list.when(
          data: (e) => Expanded(
            child: ListView.separated(
              padding: EdgeInsets.zero,
              itemBuilder: (context, index) {
                return Padding(
                  padding: const EdgeInsets.all(8.0),
                  child: Text(
                    e[index]['id'].toString(),
                  ),
                );
              },
              separatorBuilder: (context, index) {
                return const Divider();
              },
              itemCount: e.length,
            ),
          ),
          error: (obj, st) {
            return Container(
              height: 160,
              alignment: Alignment.center,
              child: ElevatedButton(
                onPressed: () => ref.invalidate(apiProvider),
                child: const Text(
                  'Retry',
                  style: TextStyle(color: Colors.white),
                  textAlign: TextAlign.center,
                ),
              ),
            );
          },
          loading: () {
            return Expanded(
              child: ListView.separated(
                itemBuilder: (context, index) =>
                    const Center(child: CircularProgressIndicator.adaptive()),
                separatorBuilder: (context, index) => const SizedBox(height: 24),
                itemCount: 4,
              ),
            );
          },
        ),
      ],
    );
  }
}
```

**Comment by AlexV525** [comment]
This is actually not caused by #2068, but a very old implementation that detaches the socket that has been promoted recently in #2036.

TL;DR, the cause is:
https://github.com/cfug/dio/blob/1c2843a2fab1c3795272df1e81502616761e192e/dio/lib/src/adapters/io_adapter.dart#L216

### Details
`ResponseBody.onClose` was implemented to detach from the socket which provided its connection in the `IOHttpClientAdapter`. However, with the same host between requests, the connections are implied to use the same socket (which is reasonable). When the `CancelToken` got canceled, it would call `ResponseBody.onClose` and the socket detached, all requests related to the same host would be closed.

The solution here might be to remove the detaching which has no affection according to its doc: https://api.dart.dev/stable/3.3.3/dart-io/HttpClientResponse/detachSocket.html
> This is normally used when a HTTP upgrade is negotiated and the communication should continue with a different protocol.

Anyway, this could be a long-affected regression, and the stack trace did not help us to combine them together, the case is rare and valunable!

**Comment by EArminjon** [comment]
What is the recommandation to avoid this issue ?

**Comment by kuhnroyal** [comment]
This was changed in 5.4.1, you might be able to lock the dio version to 5.4.0.

**Comment by devguusta** [comment]
Same problem here

**Comment by Wratheus** [comment]
Issue happens when you have two requests with different CancelTokens and while first request done, second requests starts executing and in that moment you cancel CancelToken from first request somewhere in code (for example at Bloc dispose) - then second request being canceled before was received, even if it was different cancelToken from first request whole DIO connection closes.
Fixes I found so far:
1. You can use lower Dio version.
2. Open new Dio connection for each request.
3. Initialize DIO with (BaseOptions(persistenConnection: false))
4. Ensure cancel token right after request being done in your dio controller realisation

**Comment by kuhnroyal** [comment]
@Wratheus Can you reproduce this in a test case? Also not sure if this is the same issue.

This seems to work:

```dart
    test('cancel one request', () {
      final dio = Dio()..options.baseUrl = 'https://httpbun.com';
      final token = CancelToken();

      expectLater(
        dio.get(
          '/drip?delay=0&duration=2&numbytes=10',
          cancelToken: token,
        ),
        throwsDioException(DioExceptionType.cancel),
      );
      expectLater(
        dio.get(
          '/drip?delay=0&duration=2&numbytes=10',
          cancelToken: CancelToken(),
        ),
        completes,
      );

      Future.delayed(const Duration(seconds: 1), () {
        token.cancel('cancelled');
      });
    });
```

**Comment by Wratheus** [comment]
You need to do it with post request, forgot to notice that, sorry.
This case looks alike

```dart
import 'package:dio/dio.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  test(
    'cancel one request',
    () async {
      final dio = Dio();
      const String baseUrl = 'https://httpbun.com';
      final token = CancelToken();

      await expectLater(
          dio.post(baseUrl,
              data:
                  FormData.fromMap({'delay': 0, 'duration': 1, 'numbytes': 10}),
              cancelToken: token),
          completes);
      expectLater(
        dio.post(baseUrl,
            data:
                FormData.fromMap({'delay': 0, 'duration': 5, 'numbytes': 100}),
            cancelToken: token),
        completes,
      );
      token.cancel('cancelled');
    },
  );
}

```

**Comment by AlexV525** [comment]
This case fails.

https://github.com/cfug/dio/blob/537832e17903c4046e5decc345763de3af833264/dio_test/lib/src/test/cancellation_tests.dart#L117-L143

**Comment by Wratheus** [comment]
There is no difference if there is cancelToken in second request or no.
```dart
import 'dart:async';

import 'package:dio/dio.dart';
import 'package:flutter_test/flutter_test.dart';

  void main() {
    Dio dio = Dio();
    test(
      'not closing sockets with requests that have same hosts',
          () async {
        final token = CancelToken();
        const String baseUrl = 'https://httpbun.com';

        final completer = Completer<Response?>();
        // Complete the first request with the cancel token.
        await dio.get('$baseUrl/get', cancelToken: token);
        // Request the second without any cancel token, but with the same host.
        dio.get('$baseUrl/drip?duration=3', cancelToken: CancelToken()).then(
              (res) {
            completer.complete(res);
            return res;
          },
          onError: (e, t) {
                print(e);
                print(t);
            completer.complete(null);
            return Response(requestOptions: (e as DioException).requestOptions);
          },
        );
        // Simulate connection established.
        await Future.delayed(const Duration(seconds: 1));
        token.cancel();
        final response = await completer.future;
        // Response should be obtained without exceptions.
        expect(response, isNotNull);
      },
      testOn: '!browser',
    );
  }
  ```

**Comment by chesongsong** [comment]
> 5.4.0

Is there a big difference between 5.4.0 and 5.4.2+1? I also reproduced this problem on 5.4.2+1. If I lock 5.4.0, does my user need to make API level adjustments?

**Comment by AlexV525** [comment]
The fix is included in v5.4.3 and shall not require interface updates.
